skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/SwingOffscreenDrawer.kt

```diff
@@ -0,0 +1,274 @@
+                for (line in rect.y until rect.y + rect.height) {
+                    val offset: Int = line * image.width + rect.x
+                    srcData.position(offset)[dstData, offset, min(
+                        rect.width.toDouble(),
+                        (src.capacity() - offset).toDouble()
+                    ).toInt()]
+                }
+            } else { // optimized for a buffer wide dirty rect
+                val offset: Int = rect.y * image.width
+                srcData.position(offset)[dstData, offset, min(
+                    (rect.height * image.width).toDouble(),
+                    (src.capacity() - offset).toDouble()
+                ).toInt()]
+            }
+        }
+
+        return image
+    }
+
+    private fun createVolatileImage(image: BufferedImage): VolatileImage {
+        val vi = swingLayerProperties.graphicsConfiguration.createCompatibleVolatileImage(
+            swingLayerProperties.width,
+            swingLayerProperties.height,
+            Transparency.TRANSLUCENT
+        )
+        drawVolatileImage(vi, image)
+        return vi
+    }
+
+    private fun drawVolatileImage(vi: VolatileImage, image: BufferedImage) {
+        val g = vi.graphics.create() as Graphics2D
+        try {
+            g.background = Color(0, 0, 0, 0)
+            g.composite = AlphaComposite.Src
+            g.clearRect(0, 0, swingLayerProperties.width, swingLayerProperties.height)
+            drawImage(g, image)
+        } finally {
+            g.dispose()
+        }
+    }
+
+    private fun drawImage(
+        g: Graphics,
+        image: Image,
+        x: Int = 0,
+        y: Int = 0,
+        dw: Int = -1,
+        dh: Int = -1,
+        sourceBounds: Rectangle? = null,
+        op: BufferedImageOp? = null,
+        observer: ImageObserver? = null
+    ) {
+        val hasDestinationSize = dw >= 0 && dh >= 0
+        doDrawHiDpi(
+            userWidth = swingLayerProperties.width,
+            userHeight = swingLayerProperties.height,
+            g = g,
+            scale = swingLayerProperties.scale.toDouble(),
+            dx = x,
+            dy = y,
+            dw = dw,
+            dh = dh,
+            hasDestinationSize = hasDestinationSize,
+            op = op,
+            image = image,
+            srcBounds = sourceBounds,
+            observer = observer
+        )
+    }
+
+    @Suppress("NAME_SHADOWING")
+    private fun doDrawHiDpi(
+        userWidth: Int,
+        userHeight: Int,
+        g: Graphics,
+        scale: Double,
+        dx: Int,
+        dy: Int,
+        dw: Int,
+        dh: Int,
+        hasDestinationSize: Boolean,
+        op: BufferedImageOp?,
+        image: Image,
+        srcBounds: Rectangle?,
+        observer: ImageObserver?
+    ) {
+        var g1 = g
+        var scale1 = scale
+        var dx1 = dx
+        var dy1 = dy
+        var delta = 0.0
+        // Calculate the delta based on the image size. The bigger the size - the smaller the delta.
```

Comment:
It's a good candidate to extract it to a separate function

Reply:
As written in the beginning code in this file is a copy-paste from intellij sources, so I made as few changes as possible